frame/support/src/traits.rs

```diff
@@ -1518,6 +1518,26 @@ pub trait OnFinalize<BlockNumber> {
 	fn on_finalize(_n: BlockNumber) {}
 }
 
+pub trait OnIdle<BlockNumber> {
+	/// The block is being finalized. Implement to have something happen.
+	///
+	/// NOTE: This function is called AFTER ALL extrinsics in a block are applied,
+	/// including inherent extrinsics.
+	fn on_idle(_n: BlockNumber, _remaining_weight: crate::weights::Weight) -> crate::weights::Weight { 0 }
+}
+
+#[impl_for_tuples(30)]
+impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
+	fn on_idle(_n: BlockNumber,  remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
+		let mut weight = 0;
+		for_tuples!( #(
+			let adjusted_remaining_weight = if weight < remaining_weight {remaining_weight.saturating_sub(weight)} else {0};
+			weight = weight.saturating_add(Tuple::on_idle(_n.clone(),  adjusted_remaining_weight.clone()));
```

Comment:
1. This is not doing what you want. You need to declare `adjusted_remaining_weight` outside of the closure you pass into `for_tuples` as `let mut remaining_weight = remaining_weight`. It is ok to shadow the argument.
2. No need for the `if`. This is exactly what `saturating_sub` does as long as you are dealing with unsigned types (you are).
3. Aren't those types `Copy`? I think you can drop the `clone()`. Not sure though. Just try.
4. `_n` => `n`

Reply:
actually the implementation looks fine, adjusted_remaning_weight substracted the total weight used so far.

Reply:
Then I don't get how `for_tuple` works. For me it looks like `adjusted_remaining_weight` has the same value for each call of the closure because it is defined inside the closure function rather than captured from the environment.

Reply:
value of `adjusted_remaining_weight` changes because value of `weight` is changing:
Expansion is like this:
```rust
    #[allow(unused)]
    impl <BlockNumber: Clone, TupleElement0: OnIdle<BlockNumber>,
          TupleElement1: OnIdle<BlockNumber>> OnIdle<BlockNumber> for
     (TupleElement0, TupleElement1) {
        fn on_idle(_n: BlockNumber, remaining_weight: crate::weights::Weight)
         -> crate::weights::Weight {
            let mut weight = 0;
            let adjusted_remaining_weight = if weight < remaining_weight { remaining_weight.saturating_sub(weight) } else { 0 };
            weight = weight.saturating_add(TupleElement0::on_idle(_n.clone(), adjusted_remaining_weight.clone()));
            let adjusted_remaining_weight = if weight < remaining_weight { remaining_weight.saturating_sub(weight) } else { 0 };
            weight = weight.saturating_add(TupleElement1::on_idle(_n.clone(), adjusted_remaining_weight.clone()));
            weight
        }
    }
```

Reply:
You are right. Now I get it.